### PR TITLE
Add initial pipe support to json serialization

### PR DESCRIPTION
**Pull request by steveharter**
Add PipeReader and PipeWriter support per https://github.com/dotnet/corefx/issues/34372

After the API review is completed, this will be updated again to reflect that.

cc @ahsonkhan @davidfowl

### CODE REVIEW

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.PipeReader.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                    ref readerState,
+                    returnType,
+                    result.IsCompleted,
+                    buffer,
+                    settings,
+                    ref current,
+                    ref previous,
+                    ref arrayIndex);
+
+                reader.AdvanceTo(buffer.GetPosition(readerState.BytesConsumed), buffer.End);
```
Comment (davidfowl) [inline src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.PipeReader.cs]: @ahsonkhan here's the workaround for the bug.

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.ReadOnlySequence.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+
+namespace System.Text.Json.Serialization
+{
+    public static partial class JsonConverter
+    {
+        public static T FromJson<T>(this ReadOnlySequence<byte> json, JsonConverterSettings settings = null)
```
Comment (davidfowl) [inline src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.ReadOnlySequence.cs]: Pass these by `in` I wonder how we feel about those in public APIs

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.Stream.cs [hunk 3]
```diff
@@ -15,26 +15,26 @@ public static partial class JsonConverter
     {
         private const int HalfMaxValue = int.MaxValue / 2;
 
-        public static Task<T> FromJsonAsync<T>(this Stream stream, JsonConverterSettings settings = null, CancellationToken cancellationToken = default)
+        public static Task<T> FromJsonAsync<T>(this Stream reader, JsonConverterSettings settings = null, CancellationToken cancellationToken = default)
```
Comment (davidfowl) [inline src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.Stream.cs]: Why this change?

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.ReadOnlySequence.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.ReadOnlySequence.cs]: We have it on the `Utf8JsonReader`.

**Comment by davidfowl** [comment]
Oh interesting design point:

What happens if the ReadOnlySequence\<byte\> and ReadOnlySpan\<byte\> you pass in has more data after the JSON object you de-serialized? The reason I ask is because there's no way to figure out how much of the buffer was read from the outside. 

@JamesNK what does JSON.NET do in this case?

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.Stream.cs (hunk 3 above)
Comment (steveharter) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.Stream.cs]: To align with PipeWriter and PipeReader using `writer` and `reader` naming.

**Comment by JamesNK** [comment]
https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonSerializerSettings_CheckAdditionalContent.htm

There is a setting that stops reading when serialization is complete.

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.ReadOnlySequence.cs (hunk 2 above)
Comment (steveharter) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.ReadOnlySequence.cs]: OK I'll add `in` to the ReadOnlySequence APIs

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.Stream.cs (hunk 3 above)
Comment (davidfowl) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.Stream.cs]: I don't think they should align

Comment (steveharter) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.Stream.cs]: OK I'll bring it up in the API review tomorrow.

Comment (steveharter) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.Stream.cs]: Actually using `Stream stream` is more consistent with other code, so I'll make that change now.

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.PipeReader.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Text.Json.Serialization
+{
+    public static partial class JsonConverter
+    {
+#if !BUILDING_INBOX_LIBRARY
+        [System.CLSCompliantAttribute(false)]
```
Comment (ahsonkhan) [inline src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.PipeReader.cs]: Why is this necessary?

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.PipeReader.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.PipeReader.cs]: This change should fix it: https://github.com/dotnet/corefx/pull/35245

```C#
reader.AdvanceTo(readerState.Position);
```

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.PipeReader.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                    returnType,
+                    result.IsCompleted,
+                    buffer,
+                    settings,
+                    ref current,
+                    ref previous,
+                    ref arrayIndex);
+
+                reader.AdvanceTo(buffer.GetPosition(readerState.BytesConsumed), buffer.End);
+            } while (!result.IsCompleted);
```
Comment (ahsonkhan) [inline src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.PipeReader.cs]: I think you need to pass `isFinalBlock = true` to the `Utf8JsonReader` for the last buffer.
```C#
bool isFinalBlock = result.IsCompleted;
```

Comment (ahsonkhan) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.PipeReader.cs]: Ah, I see you are doing that already (3rd parameter to `FromJson`)

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.cs [hunk 6]
```diff
@@ -16,11 +16,82 @@ public static partial class JsonConverter
 
         private static readonly JsonConverterSettings s_DefaultSettings = new JsonConverterSettings();
 
+        private static void FromJson(
+            ref JsonReaderState readerState,
+            Type returnType,
+            bool isFinalBlock,
+            byte[] buffer,
```
Comment (ahsonkhan) [inline src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.cs]: nit: Pass in ReadOnlySpan slice instead of array and int?

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.ToJson.PipeWriter.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Text.Json.Serialization
+{
+    public static partial class JsonConverter
+    {
+#if !BUILDING_INBOX_LIBRARY
+        [System.CLSCompliantAttribute(false)]
```
Comment (ahsonkhan) [inline src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.ToJson.PipeWriter.cs]: Is this causing a warning? Maybea dd a no warn exception for now (`<NoWarn>3021</NoWarn>`) since I don't think it will be needed within the source once the code is in corefx (it will only be required in the ref assembly).

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverterSettings.cs [hunk 8]
```diff
@@ -80,15 +81,23 @@ public int DefaultBufferSize
             }
             set
             {
-                if (value < 1)
+                if (value == 0 || value < BufferSizeUnspecified)
```
Comment (ahsonkhan) [inline src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverterSettings.cs]: It's odd that we allow values `<= 0`

Comment (ahsonkhan) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverterSettings.cs]: Why did we need to make this change? Now every call to `EffectiveBufferSize` has a branch.

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.cs (hunk 6 above)
Comment (steveharter) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.FromJson.cs]: This is called from an async method so it can't pass in ref structs.

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.ToJson.PipeWriter.cs (hunk 7 above)
Comment (steveharter) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.ToJson.PipeWriter.cs]: Yes this is temporary for corefxlab.

File: src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverterSettings.cs (hunk 8 above)
Comment (steveharter) [reply src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverterSettings.cs]: I didn't want to expose the actual constant (16K)
